feat: Allow setting custom api_base for OpenAI nodes

michaelfeil opened the pull request:
### Related Issues
- fixes #4923 

### Proposed Changes:
Not changing the current behaviour, by adding `api_base` similar to openai.api_base.
This makes the openai api simple to reach, in case of a proxy address.

### How did you test it?
- Ran the test locally, some tests failed that should run in the CI. 
### Notes for the reviewer
- Let me know, if I missed to implement the OpenAI api_base somewhere. 

### Checklist
- [x] I have read the [contributors guidelines](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md) and the [code of conduct](https://github.com/deepset-ai/haystack/blob/main/code_of_conduct.txt)
- [x] I have updated the related issue with new insights and changes
- [x] I added tests that demonstrate the correct behavior of the change
- [x] I've used one of the [conventional commit types](https://www.conventionalcommits.org/en/v1.0.0/) for my PR title: `fix:`, `feat:`, `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`.
- [x] I documented my code
- [x] I ran [pre-commit hooks](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md#installation) and fixed any issue

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/deepset-ai/haystack?pullRequest=5033) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/deepset-ai/haystack?pullRequest=5033) before we can accept your contribution.<br/><hr/>**michaelfeil** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/deepset-ai/haystack?pullRequest=5033) it.</sub>

michaelfeil commented:
Please give the CI another approval.

bogdankostic reviewed (changes requested):
Thanks for the PR @michaelfeil! I left a few comments regarding some doc strings and the tests. 
Also, would you be up for adding a few unit tests that test whether setting a custom base URL is working as expected?

michaelfeil commented on test/preview/components/audio/test_whisper_remote.py:
```diff
@@ -26,9 +26,10 @@ def test_init_unknown_model(self):
     @pytest.mark.unit
     def test_init_default(self):
-        transcriber = RemoteWhisperTranscriber(api_key="just a test")
+        transcriber = RemoteWhisperTranscriber(api_key="just a test", api_base="https://api.openai.com/v1")
```
Sure, this just tests the acceptance of the keyword

michaelfeil commented on test/nodes/test_retriever.py:
```diff
@@ -321,10 +321,13 @@ def test_retribert_embedding(document_store, retriever, docs_with_ids):
-    er = EmbeddingRetriever(embedding_model="text-embedding-ada-002", document_store=None)
+    er = EmbeddingRetriever(
+        embedding_model="text-embedding-ada-002", document_store=None, api_base="https://api.openai.com/v1"
+    )
```
Just changed it.

michaelfeil commented:
Added the changes as requested. I added the api_base in some unit tests, not sure how to completely mock a response without a second public proxy for openai beeing active.

michaelfeil commented:
Any chance i get permissions for the CI to run?

michaelfeil commented:
@bogdankostic Can you review this once more? Thanks in advance

dfokina commented on haystack/nodes/prompt/invocation_layer/chatgpt.py:
```diff
@@ -20,9 +20,24 @@ class ChatGPTInvocationLayer(OpenAIInvocationLayer):
+        :param api_key: The OpenAI API key.
+        :param api_base: The OpenAI API Base url, defaults to `https://api.openai.com/v1`.
+        :param kwargs: Additional keyword arguments passed to the underlying model.
+        [documentation](https://platform.openai.com/docs/api-reference/chat).
```
```suggestion
        [See OpenAI documentation](https://platform.openai.com/docs/api-reference/chat).
```

michaelfeil replied:
Okay!

bogdankostic commented on haystack/nodes/answer_generator/openai.py:
```diff
@@ -32,6 +32,7 @@ def __init__(
         api_key: str,
         azure_base_url: Optional[str] = None,
         azure_deployment_name: Optional[str] = None,
+        api_base: str = "https://api.openai.com/v1",
```
Let's move this to the end of the arguments.

bogdankostic commented on haystack/nodes/answer_generator/openai.py:
```diff
@@ -53,6 +54,7 @@ def __init__(
 
         :param azure_deployment_name: The name of the Azure OpenAI API deployment. If not supplied, Azure OpenAI API
                                      will not be used.
+        :param api_base: The base URL for the OpenAI API, defaults to `"https://api.openai.com/v1"`.
```
Let's move this to the end of doc strings given that we'll move `api_base` to the end of the arguments.

bogdankostic commented on haystack/nodes/audio/whisper_transcriber.py:
```diff
@@ -40,6 +40,7 @@ def __init__(
         self,
         api_key: Optional[str] = None,
         model_name_or_path: WhisperModel = "medium",
+        api_base: str = "https://api.openai.com/v1",
```
Let's move this to the end of the arguments.

bogdankostic commented on haystack/nodes/audio/whisper_transcriber.py:
```diff
@@ -48,12 +49,13 @@ def __init__(
         :param api_key: OpenAI API key. If None, a local installation of Whisper is used.
         :param model_name_or_path: Name of the model to use. If using a local installation of Whisper, set this to one of the following values: "tiny", "small", "medium", "large", "large-v2". If using
         the API, set thsi value to: "whisper-1" (default).
+        :param api_base: The OpenAI API Base url, defaults to `https://api.openai.com/v1`.
```
Let's move this to the end of doc strings given that we'll move `api_base` to the end of the arguments.

bogdankostic commented on haystack/nodes/retriever/dense.py:
```diff
@@ -1450,6 +1450,7 @@ def __init__(
         scale_score: bool = True,
         embed_meta_fields: Optional[List[str]] = None,
         api_key: Optional[str] = None,
+        api_base: str = "https://api.openai.com/v1",
```
Let's move this to the end of the arguments.

bogdankostic commented on haystack/nodes/retriever/dense.py:
```diff
@@ -1507,6 +1508,7 @@ def __init__(
                                   If no value is provided, a default empty list will be created.
         :param api_key: The OpenAI API key or the Cohere API key. Required if one wants to use OpenAI/Cohere embeddings.
                         For more details see https://beta.openai.com/account/api-keys and https://dashboard.cohere.ai/api-keys
+        :param api_base: The OpenAI API base URL, defaults to `"https://api.openai.com/v1"`
```
Let's move this to the end of doc strings given that we'll move `api_base` to the end of the arguments.

bogdankostic commented on haystack/preview/components/audio/whisper_remote.py:
```diff
@@ -38,11 +38,14 @@ class Input(ComponentInput):
         Transcribes a list of audio files into a list of Documents.
 
         :param api_key: OpenAI API key.
+        :param api_base: OpenAI base URL, defaults to `"https://api.openai.com/v1"`.
         :param model_name_or_path: Name of the model to use. It now accepts only `whisper-1`.
```
Let's fix the order of arguments here and while we're at it use the correct name for the `model_name` parameter.
```suggestion
        :param model_name: Name of the model to use. It now accepts only `whisper-1`.
        :param api_base: OpenAI base URL, defaults to `"https://api.openai.com/v1"`.
```

bogdankostic reviewed (changes requested):
FYI, I added some unit tests to check that the correct URL is used when a custom api base is provided.

This PR is now almost good to go, the only thing that I would like to see before merging is to move the newly introduced `api_base` parameter to the end of the arguments to not introduce a breaking change.

michaelfeil commented:
Thanks, totally missed that! Thanks for the unittest

bogdankostic reviewed (approved):
Looking good to me, thank you and congrats for your first contribution to Haystack! 🚀
